README.md
nbis
====

Fork of NIST Biometric Image Software for building and packaging on 
Linux. Upstream sources are public domain, but modifications are 
Apache-licensed.

Highlights from NBIS are:
* Fingerprint image quality assessment program (trainable)
* Fingerprint minutiae extraction program
* Fingerprint minutiae matching program
* Fingerprint Level-1 classification program (trainable)

The master branch of this repository is maintained as the code for building
packages for Debian unstable. Binary releases are available from an Ubuntu 
PPA at https://launchpad.net/~mchaberski/+archive/ppa.

The changes that were made to prepare the upstream sources for building 
as a Debian package include (but are not limited to) the following:
* the removal of imaging library source code (in favor of using imaging 
library packages in the Debian repositories);
* inclusion of static `make` rules files that were previously generated 
by a custom shell script;
* removal of obsolete files used in the upstream build system; and 
* modifications to source code to resolve runtime data in a manner that
follows conventions on Debian systems (such as reading 
architecture-independent runtime data from a system directory like 
`/usr/share/nbis`).


A more lengthy description of the contents of the NBIS distribution is below,
adapted from the [NBIS home page](http://www.nist.gov/itl/iad/ig/nbis.cfm):

> The NIST Biometric Image Software (NBIS) distribution is developed by 
the National Institute of Standards and Technology (NIST) for the 
Federal Bureau of Investigation (FBI) and Department of Homeland 
Security (DHS). The NBIS distribution contains programs for biometrics
processing and analysis. 

> The NBIS utilities included in this distribution fall under these 
general categories: 

> * A fingerprint segmentation algorithm, NFSEG, segments a
four-finger plain impression into individual fingerprint images or 
can be used to remove white space from a rolled fingerprint image.
* A neural-network based fingerprint pattern classification system, 
called PCASYS, automatically categorizes a fingerprint image into the 
class of arch, left or right loop, scar, tented arch, or whorl. This 
is an updated system that includes the use of a robust Multi-Layered 
Perceptron (MLP) neural network.
* A minutiae detector, MINDTCT, automatically locates and 
records ridge ending and bifurcations in a fingerprint image. This 
system includes minutiae quality assessment based on local image 
conditions. 
* A minutiae-based fingerprint matching algorithm, BOZORTH3, that
accepts as input minutiae locations generated by the MINDTCT 
algorithm.
* A fingerprint image quality algorithm, NFIQ, analyzes a 
fingerprint image and assigns a quality value of 1 (highest quality) 
to 5 (lowest quality) to the image. The ability to retrain the NFIQ 
weights is provided with the utilities FING2PAT, ZNORMDAT, and 
ZNORMPAT.
* A reference implementation of the ANSI/NIST-ITL 1-2007 (AN2K) 
"Data Format for the Interchange of Fingerprint, Facial, Scar Mark & 
Tattoo (SMT) Information" standard is included. This reference 
implementation contains a suite of utilities designed to read, write, 
edit, and manipulate files formatted according to this interchange 
standard. The utilities support updated and new record types 
introduced by this latest version of the standard (Record Types 9, 
13, 14, & 15).
* A large collection of general-purpose image utilities 
are also included to support the processing of fingerprint images. The 
package includes encoders and decoders for the FBI's Wavelet Scalar 
Quantization (WSQ) format. Utilities are also provided that support 
color component interleaving, colorspace conversion, and format 
conversion of legacy files distributed in NIST fingerprint databases.

As a consequence of prioritizing making the core utilities buildable 
first, the SIVV utility (spectral validation/verification metric 
for fingerprint images) is not included in packages built from this
repository.


debian/changelog
nbis (4.2.0-1) unstable; urgency=low

   * Initial release (Closes: #12)
   * added line that substitutes appropriate buildutil path into 
   p_rules.mak files
   * tilde_filename: collapsed add_path=2 option into add_path=1; i.e. 
   now tilde_filename always substitutes the installed runtime data 
   dir (if no tilde expansion is required)
   * added source code files formerly created by setup.sh (via sed 
   substitution from XXX.c.src to XXX.c)
   * added rules.mak, which was formerly generated by setup.sh; this 
   is now a static version
   * removed unnecessary imaging library directories (ijg, jpeg2k, png, 
   openjpeg)
   * generated p_rules.mak files from p_rules.mak.src, eliminating 
   need for that step in make config target
   * removed make target dependence on libopenjpeg.a and libpng.a
   * corrected openjpeg #include directive
   * replaced sed substitution with preprecessor #define for runtime 
   data directory paths (in an2k, pcasys, etc.)
   * Makefile: removed install and catalog dependencies from all 
   target divided install target into dependencies -mkdirs, 
   -runtimedata, -headers, -libnbis, -bins removed cruft relating to 
   jpeg2k, png, and openjpg packages from install-headers replaced 
   usages of FINAL_INSTALLATION_DIR with new INSTALL_ROOT_XXX_DIR 
   convention
   * rules.mak: using auto-detection of top-level directory via 
   MAKEFILE_LIST; eliminated obsolete variables; added stuff to CFLAGS
   to define runtime data directories (e.g. for an2k, pcasys); added 
   openjpeg and png libraries to LDFLAGS
   * .gitignore: added stuff that make generates to ignore list
   * Makefile: split install-mkdirs up into relevant targets; removed 
   -headers and -libnbis from install target dependencies (because 
   these would be more appropriate in a libnbis-dev package; added 
   statement to install ijg binaries that were previously missing
   * removed obsolete makefile inputs and outdated install instructions
   * modified install target to fix permissions on manual and runtime 
   data files
   * improved clean target to catch remaining *.d files and 
   bin/obj/lib directories, as well as exports and doc/catalogs
   * added apache license header to files that differ from upstream
   * moved LDFLAGS to near end of gcc command, before output 
   specification
   * added deb hardening environment variables (issue #3)
   * removed obsolete third-party makefile rules
   * fixed spelling error (completly instead of completely) 
   (issue #2); removed obsolete select.c.xxx files
   * fixed spelling error in manpages chkan2k.1 and pcasys.1 
   (issue #8)
   * removed references to jpegb (issue #4); this (unfortunately) 
   removes support for JPEG-Baseline, a popular image format, but the 
   lintian error (tag embedded-library) would have caused a build 
   failure on Launchpad
   * fixed spelling errors in chkan2k.1, pcasys.1, nfiq.c
   * added debian watch file
   * replaced bad groff_man macros with alternatives; not sure what 
   original meaning was in all cases, but this should resolve most of 
   issue #10 (except for the error regarding record_selector.1)
   * added missing man pages for dlwsqcom and optrwsgw (issue #6)
   * replaced 1[ABCDEFG] with 1 to resolve manpage section mismatches 
   (issue #5)
   * restricted man page installation to binaries that get installed 
   (instead of everything from the man/man1 source directory)
   * removed verbose echo statements from clean targets
   
 -- Michael Chaberski <mchaberski+dev@gmail.com>  Mon, 16 Dec 2013 10:28:15 -0500


setup.sh
#!/bin/sh
# ******************************************************************************
#
# License: 
# This software and/or related materials was developed at the National Institute
# of Standards and Technology (NIST) by employees of the Federal Government
# in the course of their official duties. Pursuant to title 17 Section 105
# of the United States Code, this software is not subject to copyright
# protection and is in the public domain. 
#
# This software and/or related materials have been determined to be not subject
# to the EAR (see Part 734.3 of the EAR for exact details) because it is
# a publicly available technology and software, and is freely distributed
# to any interested party with no licensing requirements.  Therefore, it is 
# permissible to distribute this software as a free download from the internet.
#
# Disclaimer: 
# This software and/or related materials was developed to promote biometric
# standards and biometric technology testing for the Federal Government
# in accordance with the USA PATRIOT Act and the Enhanced Border Security
# and Visa Entry Reform Act. Specific hardware and software products identified
# in this software were used in order to perform the software development.
# In no case does such identification imply recommendation or endorsement
# by the National Institute of Standards and Technology, nor does it imply that
# the products and equipment identified are necessarily the best available
# for the purpose.
#
# This software and/or related materials are provided "AS-IS" without warranty
# of any kind including NO WARRANTY OF PERFORMANCE, MERCHANTABILITY,
# NO WARRANTY OF NON-INFRINGEMENT OF ANY 3RD PARTY INTELLECTUAL PROPERTY
# or FITNESS FOR A PARTICULAR PURPOSE or for any purpose whatsoever, for the
# licensed product, however used. In no event shall NIST be liable for any
# damages and/or costs, including but not limited to incidental or consequential
# damages of any kind, including economic damage or injury to property and lost
# profits, regardless of whether NIST shall be advised, have reason to know,
# or in fact shall know of the possibility.
#
# By using this software, you agree to bear all risk relating to quality,
# use and performance of the software and/or related materials.  You agree
# to hold the Government harmless from any claim arising from your use
# of the software.
#
#
# ******************************************************************************
# Project:              NIST Fingerprint Software
# SubTree:              /NBIS/Main
# Filename:             setup.sh
# Integrators:          Kenneth Ko
# Organization:         NIST/ITL
# Host System:          GNU GCC/GMAKE GENERIC (UNIX)
# Date Created:         08/20/2006
# Date Updated:         02/20/2007 (Kenneth Ko)
#                       10/23/2007 (Kenneth Ko)
#                       11/16/2007 (Kenneth Ko)
#                       01/31/2008 (Kenneth Ko)
#                       03/20/2008 (Kenneth Ko)
#                       03/27/2008 (Kenneth Ko)
#                       04/14/2008 (Kenneth Ko)
#                       09/03/2008 (Kenneth Ko)
#                       12/02/2008 (Kenneth Ko) - Fix to support 64-bit
#                       12/16/2008 (Kenneth Ko) - Add command line option for
#                                                 HPUX.
#                       12/17/2008 (Kenneth Ko) - Reorganize what packages to 
#                                                 use to build the software on
#                                                 HPUX.
#                       01/16/2009 (Greg Fiumara) - Switch standard library
#                                                   configuration to 
#                                                   __NBIS_<STDLIBS>__
#                       01/09/2009 (Greg Fiumara) - Add --64/--32 option to 
#                                                   force architecture on
#                                                   cross-compile capable 
#                                                   systems.
#                       05/02/2011 (Kenneth Ko) - Update licensing information.
#                                                 Add fPIC flag.
#                                                 Build PNG libaray as default 
#                                                 on MSYS.
#                       05/09/2012 (Kenneth Ko) - Remove 32-bit or 64-bit 
#                                                 architecture check.
#                                                 Default to build with fPIC.
#                                                 Remove --with-JASPER support.
#                                                 Add --without-OPENJPEG option
#                                                 to disable building OpenJPEG 
#                                                 library.
#                                                 Check arch.mak exist or not 
#                                                 before delete.
#                       10/22/2013 (Kenneth Ko) - Bug fix for MSYS installation. 
#                       
#                                                  
# ******************************************************************************
#
# This is a setup script. Run this prior to compilation.
#
# ******************************************************************************
# Check number of argument
if [ $# -lt 1 -o $# -gt 4 ] ; then 
	# ERROR
	echo "    For detail information on how to build NBIS, please consult the";
	echo "    INSTALLATION_GUIDE.";
	exit 1
fi

# Check the first argument is directory
if [ ! -d $1 ]; then
	# ERROR
	echo "Directory \"$1\" doesn't exist!";
	exit 1
else
	FINAL_INSTALLATION_DIR=$1
	INSTALL_DATA_DIR_STRING=$FINAL_INSTALLATION_DIR/nbis
	shift
fi

#--------------------------- Check and Set Flag --------------------------------
X11_FLAG=1
STDLIBS_FLAG=0
NBIS_JASPER_FLAG=0
NBIS_PNG_FLAG=1
NBIS_OPENJPEG_FLAG=1
X86_64_FLAG=-1
TARGET_OS=0
MSYS_FLAG=0
FPIC_FLAG=0

while [ $# -ge 1 ]; do
	case $1 in
	--without-X11)
		X11_FLAG=0;;
	--STDLIBS)
		STDLIBS_FLAG=1
		NBIS_PNG_FLAG=0
		NBIS_OPENJPEG_FLAG=0;;
	--without-OPENJPEG)
		NBIS_OPENJPEG_FLAG=0;;
	--64)         
		X86_64_FLAG=1;;
	--32)         
		X86_64_FLAG=0;;
	--SUPERDOME)
		TARGET_OS=1
		STDLIBS_FLAG=1
		NBIS_PNG_FLAG=0
		NBIS_OPENJPEG_FLAG=0
		X11_FLAG=0;;
	--MSYS)
		MSYS_FLAG=1
		STDLIBS_FLAG=1
		X11_FLAG=0;;
	*)
		echo "No such option: $1"
		exit 1;;
	esac
	shift
done

#---------------------------------- X11 path -----------------------------------
OS_FLAG=0

if [ `uname` = "Linux" ]; then
	X11_INC="/usr"
        X11_LIB="/usr"
elif [ `uname` = "Darwin" ]; then  
	X11_INC="/usr/X11R6/include"
	X11_LIB="/usr/X11R6/lib"
	OS_FLAG=1
elif [ `uname` = "CYGWIN_NT-5.1" ]; then
	X11_INC="/usr/X11R6/include"
	X11_LIB="/usr/X11R6/lib"
else
	X11_INC="/usr/X11R6"
        X11_LIB="/usr/X11R6"
fi

if [ $X11_FLAG -eq 1 ] ; then
	if [ ! -d $X11_INC ]; then
		echo "Directory $X11_INC doesn't exist!";
		echo "Please update variable X11_INC in setup.sh, if you wish to compile with X11."
		exit 1
	fi
	if [ ! -d $X11_LIB ]; then
		echo "Library $X11_LIB not find!";
		echo "Please update variable X11_INC in setup.sh, if you wish to compile with X11."
		exit 1
	fi
fi

#------------------------------- Set Current Path --------------------------------
# Set Path
if [ $MSYS_FLAG -eq 1 ]; then
        MAIN_DIR=$PWD
        cd ..
        NBIS_DIR=`dirname $PWD;`
else
        MAIN_DIR=$PWD
        cd ..
        NBIS_DIR=$PWD
fi

cd $MAIN_DIR

#---------------------------------- Check Endian ---------------------------------
# Detect target machine OS
rulesFile="rules.mak.src"

# Detect target machine's endian
if [ $MSYS_FLAG -eq 1 ]; then
	RM=rm
	CC=gcc
	CP=cp
else
	RM=`which rm`
	CC=gcc
	CP=`which cp`
fi

endianFile=./endian.out

if [ -e ${endianFile} ]; then
	`${RM} ${endianFile}`
fi

`${CC} am_big_endian.c -o am_big_endian`
`./am_big_endian > ${endianFile}`

if [ ! -e ${endianFile} ]; then
	echo "Failed Setup - Cannot find file - ${endianFile}"
	exit 1
fi

if [ ! -s ${endianFile} ]; then
	echo "Failed Setup - Cannot determine target machine endianness!"
	exit 1  
fi

endian=`head -n 1 ${endianFile}`

# See if the compiler supports generating position independent code
FPIC_FLAG=1

#----------------------------- Use "sed" to Modify ------------------------------
# Set PACKAGES varabile in rules.mak.src which depend on the export control
# directory(s) exist or not.
IJG='ijg'
JPEG2K='jpeg2k'
NBIS_PNG='png'
OPENJPEG='openjpeg'
REG_LIBS='commonnbis an2k bozorth3 imgtools mindtct nfseg nfiq pcasys'

if [ $NBIS_JASPER_FLAG -eq 1 -a $NBIS_PNG_FLAG -eq 0 ]; then
	IMAGE_LIBS="${IJG} ${JPEG2K}"
elif [ $NBIS_JASPER_FLAG -eq 0 -a $NBIS_PNG_FLAG -eq 1 ]; then
	IMAGE_LIBS="${IJG} ${NBIS_PNG}"
elif [ $NBIS_JASPER_FLAG -eq 1 -a $NBIS_PNG_FLAG -eq 1 ]; then
	IMAGE_LIBS="${IJG} ${JPEG2K} ${NBIS_PNG}"
elif [ $NBIS_JASPER_FLAG -eq 0 -a $NBIS_PNG_FLAG -eq 0 ]; then
	IMAGE_LIBS="${IJG}"
fi

if [ $NBIS_OPENJPEG_FLAG -eq 1 ]; then
	IMAGE_LIBS="${IMAGE_LIBS} ${OPENJPEG}"
fi

PACKAGES="${IMAGE_LIBS} ${REG_LIBS}"

if [ $MSYS_FLAG -eq 1 ]; then 
	cat rules_msys.mak.src | sed 's,SED_PACKAGES_STRING,'"$PACKAGES"',' > rules.mak.temp1
elif [ $TARGET_OS -eq 1 ]; then 
	cat rules_superdome.mak.src | sed 's,SED_PACKAGES_STRING,'"$PACKAGES"',' > rules.mak.temp1
else
	cat rules.mak.src | sed 's,SED_PACKAGES_STRING,'"$PACKAGES"',' > rules.mak.temp1
fi

# Use 'sed' command to modify all the necessary files
cat rules.mak.temp1 | sed 's,SED_DIR_MAIN,'$MAIN_DIR',' > rules.mak.temp2
cat rules.mak.temp2 | sed 's,SED_FINAL_INSTALLATION_STRING,'$FINAL_INSTALLATION_DIR',' > rules.mak.temp3
cat rules.mak.temp3 | sed 's,SED_X11_FLAG,'$X11_FLAG',' > rules.mak.temp4
cat rules.mak.temp4 | sed 's,SED_X11_INC,'$X11_INC',' > rules.mak.temp5
cat rules.mak.temp5 | sed 's,SED_X11_LIB,'$X11_LIB',' > rules.mak.temp6

# Use 'sed' command to modify ENDIAN_FLAG
if [ $endian -eq 1 ]; then
	ENDIAN_FLAG="-D__NBISLE__"
elif [ $endian -eq 0 ]; then
	ENDIAN_FLAG="";
fi
cat rules.mak.temp6 | sed 's,SED_ENDIAN_FLAG,'$ENDIAN_FLAG',' > rules.mak.temp7

# Use 'sed' command to modify J_FLAG
if [ $NBIS_JASPER_FLAG -eq 1 ]; then
	J_FLAG="-D__NBIS_JASPER__"
	cat rules.mak.temp7 | sed 's,SED_NBIS_JASPER_FLAG,'$J_FLAG',' > rules.mak.temp8
elif [ $NBIS_JASPER_FLAG -eq 0 ]; then
	J_FLAG="";
	cat rules.mak.temp7 | sed 's,SED_NBIS_JASPER_FLAG,'$J_FLAG',' > rules.mak.temp8
fi

# Use 'sed' command to modify ARCH_FLAG
if [ -e arch.mak ]; then
	`${RM} arch.mak`
fi

if [ $X86_64_FLAG -eq -1 ]; then
	ARCH_FLAG=""
	echo "ARCH_FLAG = " >> arch.mak
elif [ $X86_64_FLAG -eq 1 ]; then
	ARCH_FLAG="-m64"
	echo "ARCH_FLAG = -m64" >> arch.mak
elif [ $X86_64_FLAG -eq 0 ]; then
	ARCH_FLAG="-m32"
	echo "ARCH_FLAG = -m32" >> arch.mak
fi
if [ $FPIC_FLAG -eq 1 ]; then
	ARCH_FLAG="$ARCH_FLAG -fPIC"
fi
echo "ARCH_FLAG = $ARCH_FLAG" >> arch.mak
cat rules.mak.temp8 | sed "s,SED_ARCH_FLAG,$ARCH_FLAG," > rules.mak.temp9

# Use 'sed' command to modify P_FLAG
if [ $NBIS_PNG_FLAG -eq 1 ]; then
        P_FLAG="-D__NBIS_PNG__"
        cat rules.mak.temp9 | sed 's,SED_NBIS_PNG_FLAG,'$P_FLAG',' > rules.mak.temp10
elif [ $NBIS_PNG_FLAG -eq 0 ]; then
        P_FLAG="";
        cat rules.mak.temp9 | sed 's,SED_NBIS_PNG_FLAG,'$P_FLAG',' > rules.mak.temp10
fi

# Use 'sed' commdand to modify MSYS_FLAG
if [ $MSYS_FLAG -eq 1 ]; then
	M_FLAG="-D__MSYS__"
	cat rules.mak.temp10 | sed 's,SED_MSYS_FLAG,'$M_FLAG',' > rules.mak.temp11
elif [ $MSYS_FLAG -eq 0 ]; then
	M_FLAG=""
	cat rules.mak.temp10 | sed 's,SED_MSYS_FLAG,'$M_FLAG',' > rules.mak.temp11
fi

# Use 'sed' commdand to modify OS_FLAG
if [ $OS_FLAG -eq 1 ]; then
	O_FLAG="DARWIN"
	cat rules.mak.temp11 | sed 's,SED_OS_FLAG,'$O_FLAG',' > rules.mak.temp12
elif [ $OS_FLAG -eq 0 ]; then
	O_FLAG=""
	cat rules.mak.temp11 | sed 's,SED_OS_FLAG,'$O_FLAG',' > rules.mak.temp12
fi

# Use 'sed' commdand to modify NBIS_OPENJPEG_FLAG
if [ $NBIS_OPENJPEG_FLAG -eq 1 ]; then
	O_FLAG="-D__NBIS_OPENJPEG__"
	cat rules.mak.temp12 | sed 's,SED_NBIS_OPENJPEG_FLAG,'$O_FLAG',' > rules.mak
elif [ $NBIS_OPENJPEG_FLAG -eq 0 ]; then
	O_FLAG="";
	cat rules.mak.temp12 | sed 's,SED_NBIS_OPENJPEG_FLAG,'$O_FLAG',' > rules.mak
fi

rm -f rules.mak.temp1 rules.mak.temp2 rules.mak.temp3 rules.mak.temp4 rules.mak.temp5 rules.mak.temp6 rules.mak.temp7 rules.mak.temp8 rules.mak.temp9 rules.mak.temp10 rules.mak.temp11 rules.mak.temp12

cd ${MAIN_DIR}/an2k/include
sed 's,SED_INSTALL_DATA_DIR_STRING,'$INSTALL_DATA_DIR_STRING',' < an2k.h.src > an2k.h

cd ${MAIN_DIR}/pcasys/include
cat little.h.src | sed 's,SED_INSTALL_DIR_STRING,'$MAIN_DIR',' > little.h.temp1
cat little.h.temp1 | sed 's,SED_INSTALL_DATA_DIR_STRING,'$INSTALL_DATA_DIR_STRING',' > little.h.temp2
cat little.h.temp2 | sed 's,SED_INSTALL_NBIS_DIR_STRING,'$NBIS_DIR',' > little.h
rm -f little.h.temp1 little.h.temp2

cd ${MAIN_DIR}/pcasys/src/bin/optrws
cat optrws.c.src | sed 's,SED_INSTALL_BIN_DIR_STRING,'$INSTALL_BIN_DIR_STRING',' > optrws.c

if [ $TARGET_OS -eq 1 ]; then
	cd ${MAIN_DIR}/an2k/src/lib/an2k
	${CP} select.c.superdome select.c
else
	cd ${MAIN_DIR}/an2k/src/lib/an2k
	${CP} select.c.linux select.c
fi

chmod +w ${MAIN_DIR}/jpeg2k/src/lib/jasper/configure
chmod +w ${MAIN_DIR}/jpeg2k/src/lib/jasper/aclocal.m4
chmod +w ${MAIN_DIR}//an2k/src/lib/an2k/select.c
